From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kuan-Chih Wang <kuanchihwang@outlook.com>
Date: Sun, 3 May 2026 09:11:01 +0800
Subject: [PATCH] Fix feature detection logic for MPI-related capabilities

The `FindMPI` CMake module already sets `MPI_Fortran_HAVE_F77_HEADER` and
`MPI_Fortran_HAVE_F90_MODULE`. Just use them and stop improvising broken ways
to attempt the same thing but unsuccessfully.
---
 src/flib/CMakeLists.txt | 25 ++++++-------------------
 src/gptl/CMakeLists.txt | 10 +---------
 2 files changed, 7 insertions(+), 28 deletions(-)

diff --git a/src/flib/CMakeLists.txt b/src/flib/CMakeLists.txt
index 1562b29..6adac28 100644
--- a/src/flib/CMakeLists.txt
+++ b/src/flib/CMakeLists.txt
@@ -188,19 +188,14 @@ if(PIO_USE_MPISERIAL)
   endif()
 endif()
 
-# Check MPI I/O capabilities
-find_path(MPIF_H_PATH
-  NAMES mpif.h
-  PATHS ${MPI_Fortran_INCLUDE_PATH}
-  NO_DEFAULT_PATH)
+# Check for MPI I/O capabilities
 
-if(MPIF_H_PATH)
+if(MPI_Fortran_HAVE_F90_MODULE)
   check_macro(MPI_HAS_MPIIO
     NAME TryMPIIO.f90
     HINTS ${CMAKE_MODULE_PATH}
-    DEFINITIONS -I${MPIF_H_PATH}
     COMMENT "whether MPIIO is supported")
-  if(${MPI_HAS_MPIIO})
+  if(MPI_HAS_MPIIO)
     message(STATUS "MPIIO verified and enabled.")
     target_compile_definitions(piof
       PUBLIC USEMPIIO)
@@ -213,16 +208,7 @@ endif()
 
 # Check for MPI Fortran module
 
-find_path(MPIMOD_PATH
-  NAMES mpi.mod MPI.mod
-  HINTS ${MPI_Fortran_INCLUDE_PATH})
-
-check_macro(MPI_HAS_Fortran_MOD
-  NAME TryMPIMod.f90
-  HINTS ${CMAKE_MODULE_PATH}
-  DEFINITIONS -I${MPIMOD_PATH}
-  COMMENT "whether MPI Fortran module is supported")
-if(${MPI_HAS_Fortran_MOD})
+if(MPI_Fortran_HAVE_F90_MODULE)
   message(STATUS "MPI Fortran module verified and enabled.")
 else()
   message(STATUS "MPI Fortran module failed verification and therefore disabled.")
@@ -230,7 +216,8 @@ else()
     target_compile_definitions(gptl
       PUBLIC NO_MPIMOD)
   endif()
-  target_compile_definitions(piof PUBLIC NO_MPIMOD)
+  target_compile_definitions(piof
+    PUBLIC NO_MPIMOD)
 endif()
 
 #===== GPTL =====
diff --git a/src/gptl/CMakeLists.txt b/src/gptl/CMakeLists.txt
index 83dca72..6151bfd 100644
--- a/src/gptl/CMakeLists.txt
+++ b/src/gptl/CMakeLists.txt
@@ -151,15 +151,7 @@ endif()
 
 # Check for MPI Fortran module
 
-find_path(MPIMOD_PATH
-  NAMES mpi.mod MPI.mod
-  PATHS ${MPI_Fortran_INCLUDE_PATH})
-check_macro(MPI_HAS_Fortran_MOD
-  NAME TryMPIMod.f90
-  HINTS ${CMAKE_MODULE_PATH}
-  DEFINITIONS -I${MPIMOD_PATH}
-  COMMENT "whether MPI Fortran module is supported")
-if(${MPI_HAS_Fortran_MOD})
+if(MPI_Fortran_HAVE_F90_MODULE)
   message(STATUS "MPI Fortran module verified and enabled.")
 else()
   message(STATUS "MPI Fortran module failed verification and therefore disabled.")
-- 
2.47.3

